### ISSUE TITLE
torch.compile silently incorrect on double backwards

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#91469]
## Code example

The second-order gradients are completely bogus:
```
def f(x):
    # Just to make sure there isn't a limit or something
    y = x * x * x * x * x * 1 * 1 * 1. * 1. * 1.
    print("graph_break")
    gx, = torch.autograd.grad(y, x, create_graph=True)
    ggx, = torch.autograd.grad(gx, x)
    return gx, ggx

x = torch.tensor(1., requires_grad=True)
of = torch.compile(backend='aot_eager')(f)

gx, ggx = f(x)
ogx, oggx = of(x)
assert torch.allclose(gx, ogx)
# BUG: second-order gradients are completely bogus. assertion failed
assert torch.allclose(ggx, oggx)
```

**Comment by ngimel** [comment]
High priority for silently incorrect results

**Comment by albanD** [comment]
I think there are two cases we care about here btw:
- The compiled function does not contain call to backward but is differentiated through twice
- The compiled function does contain call(s) to backward

Both of these should be nicely disabled.

**Comment by zou3519** [comment]
> I think there are two cases we care about here btw:
> 
> * The compiled function does not contain call to backward but is differentiated through twice
> * The compiled function does contain call(s) to backward
> 
> Both of these should be nicely disabled.

Agree that both of these should either work or error out with a nice error message. I tested both (though the issue only mentions one) -- both are silently incorrect.

**Comment by soulitzer** [comment]
https://github.com/pytorch/pytorch/pull/92348 has landed, so this is no longer a silent correctness issue, updated the title to reflect that.

**Comment by milmor** [comment]
torch.compile is awesome, but I can't harness it because it's not possible to train GANs. Tensorflow GAN with jit=True trains 2x faster than torch without compile. This is sad :( I love pytorch <3

**Comment by ezyang** [comment]
FWIW, if your model is `make_fx` traceable, you can compile with double backwards by just bypassing aotautograd and tracing the consolidate forward-backward-double backwards. But yeah, we should just fix this... it's kind of a pain though.

**Comment by RaulPPelaez** [comment]
Hi! I wanted to bring some attention to this feature. Our models, in TorchMD-Net https://github.com/torchmd/torchmd-net/issues/307 , call autograd as part of forward, preventing us from training using torch.compile.

In essence, our models are set up like this:
```python
import torch
from torch import nn, Tensor
from typing import Tuple

class Model(nn.Module):
    def forward(self, input: Tensor) -> Tuple[Tensor, Tensor]:
        output = input * input # Arbitrary operation
        grad_output = torch.autograd.grad(
            output,
            input,
            grad_outputs=torch.ones_like(output),
            create_graph=True,
            retain_graph=True,
        )[0]
        return output, grad_output

model = Model()
model = torch.compile(model, backend="inductor")
input = torch.randn(10, requires_grad=True)
y, dy = model(input)
# Training with torch.compile would require this line to be possible.
ddy = torch.autograd.grad(dy, input, grad_outputs=torch.ones_like(dy))[0]

```
Even with the latest nightly the above code produces the message that titles this issue: `torch.compile with aotautograd does not support double backwards`
`pytorch                   2.3.0.dev20240313    py3.11_cpu_0    pytorch-nightly`

cc @chauhang

**Comment by ezyang** [comment]
We have a plan which should in principle solve this problem https://docs.google.com/document/d/1EHEMwSzBPOsmk0EDztRXlGwEo_TizVc5jkW_5Gp5eRs/edit which @bdhirsh is planning to work on.

**Comment by zou3519** [comment]
See https://github.com/pytorch/pytorch/issues/91542 for some more context and motivation

**Comment by bdhirsh** [comment]
Hoping to start on this in a few weeks. Just to lay out some of the things that will be fixed by the current [rough design](https://docs.google.com/document/d/1EHEMwSzBPOsmk0EDztRXlGwEo_TizVc5jkW_5Gp5eRs/edit#heading=h.moip2jp9ghoz)

(1) double backward enablement in compile

(2) avoid raising an error today when we incorrectly guess the metadata of tensor subclass tangents in the backward

(3) fix compiled autograd support for tensor subclasses at graph boundaries

A few of the challenges:

(1) need to move backward compilation to always happen in the backward. Right now, if we have any symints saved-for-backward, we do the backward compilation ahead of time. Moving it to the backward will require handling the case where inductor decides to add any extra guards on these backward symints (either by asking inductor to tell us ahead-of-time what these guards will be, or adding an extra dynamic dispatch layer in the backward)

(2) need to figure out tracing the backward section of the joint graph in pre_dispatch, including putting tensor subclass contructors in the graph

(3) need to make the partitioner happy with the backward being predispatch

(4) need to make warm cache happy

**Comment by YutackPark** [comment]
Hi, similar to above TorchMD-Net, our model calls `torch.autograd.grad` internally. I was unable to use `torch.compile` for training (I got an assertion error, and found this issue). But somehow, it raises no error in this nightly version.

'2.6.0.dev20241112+cu121'

Mine is slightly different. I call `torch.autograd.grad` then `loss.backward()`. but it should be relevant to this issue. 
Is it solved in nightly version? Sorry for the impatient. I really want torch.compile.

**Comment by caic99** [comment]
@YutackPark Maybe you could try the approaches from https://arxiv.org/pdf/2504.16068
